simpeg/directives/_regularization.py

```diff
@@ -0,0 +1,482 @@
+from __future__ import annotations
+
+import warnings
+
+import numpy as np
+from dataclasses import dataclass
+from typing import Literal
+from .directives import InversionDirective, UpdatePreconditioner, BetaSchedule
+from simpeg.regularization import Sparse, BaseSparse, SmoothnessFirstOrder
+from simpeg.utils import validate_integer, validate_float
+from simpeg.utils.code_utils import deprecate_property
+
+
+@dataclass
+class IRLSMetrics:
+    """
+    Data class to store metrics used by the IRLS algorithm.
+
+    Parameters
+    ----------
+
+    input_norms: List of norms temporarily stored during the initialization.
+    irls_iteration_count: Number of IRLS iterations.
+    start_irls_iter: Iteration number when the IRLS process started.
+    f_old: Previous value of the regularization function.
+    mode: Mode of the IRLS algorithm, 1 for l2-norm and 2 for lp-norm iterations.
+    """
+
+    input_norms: list[float]
+    irls_iteration_count = 0
+    start_irls_iter = 0
+    f_old: float = 0.0
+    mode: Literal[1, 2] = 1
+
+
+class UpdateIRLS(InversionDirective):
+    """
+    Directive to control the IRLS iterations for regularization.Sparse.
+
+    Parameters
+    ----------
+
+    beta_search: bool
+        Proceed with a beta search step if the target misfit is outside the tolerance.
+    misfit_tolerance: float
+        Tolerance for the target misfit.
+    chifact_start: float
+        Starting chi factor for the IRLS iterations.
+    chifact_target: float
+        Target chi factor for the IRLS iterations.
+    cooling_factor: float
+        Factor to cool the IRLS threshold epsilon.
+    f_min_change: float
+        Minimum change in the regularization function to continue the IRLS iterations.
+    max_irls_iterations: int
+        Maximum number of IRLS iterations.
+    percentile: float
+        Percentile of the function values used to determine the initial IRLS threshold.
+    """
+
+    def __init__(
+        self,
+        beta_schedule: BetaSchedule | None = None,
+        chifact_start: float = 1.0,
+        chifact_target: float = 1.0,
+        cooling_factor: float = 1.2,
+        f_min_change: float = 1e-2,
+        max_irls_iterations: int = 20,
+        misfit_tolerance: float = 1e-1,
+        percentile: int = 100,
```

Comment:
Updating the type hint according to the docstring. Is it right?
```suggestion
        percentile: float = 100,
```